### PR TITLE
Docs and some missing functions for chai.assert

### LINKED ISSUE #14399: chai definitions should use generics instead of any
- [x] I tried using the `@types/xxxx` package and had problems.

For example, `equal` is defined as:
```typescript
equal(act: any, exp: any, msg?: string): void;
```

IMO it should be:
```typescript
equal<T>(actual: T, expected: T, message?: string): void;
```

I'm willing to make this large sweeping change, if that helps.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Bumped Chai version to 3.5.0.

Fixes #14399.

### CODE REVIEW

**Comment by JoshuaKGoldberg** [comment]
It doesn't pass Travis CI because of TypeScript 2.1 - specific features... but I did as the DT README.md [suggests](https://github.com/DefinitelyTyped/DefinitelyTyped/#i-want-to-use-features-from-typescript-21) and added the 2.1 comment. This is confusing.

**Comment by mhegazy** [comment]
Chai has multiple dependencies; by pumping it to TS 2.1, you need to update them as well.

**Comment by JoshuaKGoldberg** [comment]
@mhegazy is this requirement documented anywhere?

**Comment by JoshuaKGoldberg** [comment]
After a second glance, I don't follow. Why does Chai having a dependency on something necessitate it being bumped to 2.1? 

On the flip side, does bumping Chai to 2.1 mean anything that depends on it should also be bumped to 2.1? That would make more sense, but probably be a huge change. If this is the case, it might make sense to file an issue to generate a tool that auto-bumps everything depending on a 2.1-requiring typing?

**Comment by paulvanbrenk** [comment]
It's the 2nd case.. once you bump Chai, everything that depends on it needs to be bumped (and potentially fixed) as well... please file a bug so we can get some traction on making that process easier.

**Comment by JoshuaKGoldberg** [comment]
Alternative proposal: I can remove the TS 2.1-specific features. The only blocker was a string literal type. Could you please re-open the PR @paulvanbrenk? I don't think I have permissions to.

**Comment by paulvanbrenk** [comment]
@JoshuaKGoldberg  done

**Comment by JoshuaKGoldberg** [comment]
Thanks!

Re-pinging authors (@jedmao @.Bartvds @AGBrown @olivr70 @mwistrand). Could you review this PR?
👍 or 👎?

File: types/chai/index.d.ts [hunk 1]
```diff
@@ -17,14 +18,15 @@ declare namespace Chai {
         /**
          * Provides a way to extend the internals of Chai
          */
-        use(fn: (chai: any, utils: any) => void): ChaiStatic;
+        use(fn: (chai: (any), utils: any) => void): ChaiStatic;
```
Comment (jedmao) [inline types/chai/index.d.ts]: Are these parentheses adding any value?

File: types/chai/index.d.ts [hunk 2]
```diff
@@ -34,6 +36,10 @@ declare namespace Chai {
         (target: any, message?: string): Assertion;
     }
 
+    export type Operator = string; // "==" | "===" | ">" | ">=" | "<" | "<=" | "!=" | "!==";
```
Comment (jedmao) [inline types/chai/index.d.ts]: I think you can use the actual values here:
```ts
export type Operator = '==' | '===' | '>' | '>=' | '<' | '<=' | '!=' | '!==';
```

File: types/chai/index.d.ts [hunk 3]
```diff
@@ -406,12 +1241,12 @@ declare namespace Chai {
...
+declare const chai: Chai.ChaiStatic;
 
 declare module "chai" {
     export = chai;
 }
 
 interface Object {
     should: Chai.Assertion;
-}
+}
```
Comment (jedmao) [inline types/chai/index.d.ts]: nope

Comment (jedmao) [inline types/chai/index.d.ts]: nope

**Review by jednano — CHANGES_REQUESTED** [review]
It's looking pretty good. Just a couple minor comments.

File: types/chai/index.d.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [inline types/chai/index.d.ts]: I honestly have no idea where they came from. Will remove...

File: types/chai/index.d.ts (hunk 2 above)
Comment (JoshuaKGoldberg) [inline types/chai/index.d.ts]: Can't 😦 see above conversation. This and the `keyof` types are blocked on updating libraries that depend on chai to also require TS 2.1. I'll do that in a subsequent PR.

File: types/chai/index.d.ts (hunk 3 above)
Comment (JoshuaKGoldberg) [inline types/chai/index.d.ts]: Will fix!

**Review by jednano — APPROVED** [review]

File: types/chai/index.d.ts [hunk 4]
```diff
@@ -238,146 +244,975 @@ declare namespace Chai {
...
+         * string using dot- and bracket-notation for deep reference.
+         * 
+         * @type T   Type of object.
+         * @type V   Type of value.
+         * @param object   Container object.
+         * @param property   Potential contained property of object.
+         * @param value   Potential expected property value.
+         * @param message   Message to display on error.
+         */
+        deepPropertyNotVal<T, V>(object: T, property: string, value: T, message?: string): void;
```
Comment (Agamnentzar) [inline types/chai/index.d.ts]: shouldn't this be `value: V` ?

Comment (JoshuaKGoldberg) [inline types/chai/index.d.ts]: Oh good catch, thank you! Will file a followup PR.

**Pull request merged**